### PR TITLE
Defined onchange of track_all to check whether there exists any quants for this product without serial number to avoid reserveing of quants without lot in trazeable products

### PR DESCRIPTION
**Pull request by dhr-odoo**

### CODE REVIEW

File: addons/stock/product.py [hunk 1]
```diff
@@ -287,6 +287,19 @@ def action_view_routes(self, cr, uid, ids, context=None):
         template_obj = self.pool.get("product.template")
         templ_ids = list(set([x.product_tmpl_id.id for x in self.browse(cr, uid, ids, context=context)]))
         return template_obj.action_view_routes(cr, uid, templ_ids, context=context)
+    
+    def onchange_track_all(self, cr, uid, ids, track_all, context=None):
+        warning = {}
+        if ids and track_all:
+            quant_ids = self.pool.get('stock.quant').search(cr, uid, [('product_id','=',ids), ('lot_id','=', False), ('location_id.usage','=', 'internal')])
```
Comment (jco-odoo) [inline addons/stock/product.py]: Here you should call the function for the related template as the track_all is defined on product template level.  self.pool.get('product.template').onchange_track_all (cr, uid, [...  (ids of product_tmpl_id)

File: addons/stock/product.py [hunk 2]
```diff
@@ -402,6 +415,18 @@ def action_view_routes(self, cr, uid, ids, context=None):
         result['domain'] = "[('id','in',[" + ','.join(map(str, route_ids)) + "])]"
         return result
 
+    def onchange_track_all(self, cr, uid, ids, track_all, context=None):
+        warning = {}
+        if ids and track_all:
+            quant_ids = self.pool.get('stock.quant').search(cr, uid, [('product_id.product_tmpl_id','in',ids), ('lot_id','=', False), ('location_id.usage','=', 'internal')])
+            if quant_ids:
+                warning_msgs =  _("There are Quants without lot for this product. You need to serialize them first to prevent inventory loss.")
```
Comment (jco-odoo) [inline addons/stock/product.py]: I don't think you would use 'serialize' in this case.  I would prefer 'assign a serial number (e.g. by creating an inventory)'

File: addons/stock/product.py [hunk 3]
```diff
@@ -402,6 +406,18 @@ def action_view_routes(self, cr, uid, ids, context=None):
         result['domain'] = "[('id','in',[" + ','.join(map(str, route_ids)) + "])]"
         return result
 
+    def onchange_track_all(self, cr, uid, ids, track_all, context=None):
+        warning = {}
+        if ids and track_all:
+            quant_ids = self.pool.get('stock.quant').search(cr, uid, [('product_id.product_tmpl_id','in',ids), ('lot_id','=', False), ('location_id.usage','=', 'internal')])
```
Comment (rim-odoo) [inline addons/stock/product.py]: @dhr-odoo Use `count=True`, as you will never read `quant_ids`, you just want to check they exist

File: addons/stock/product.py [hunk 4]
```diff
@@ -402,6 +406,18 @@ def action_view_routes(self, cr, uid, ids, context=None):
         result['domain'] = "[('id','in',[" + ','.join(map(str, route_ids)) + "])]"
         return result
 
+    def onchange_track_all(self, cr, uid, ids, track_all, context=None):
+        warning = {}
```
Comment (rim-odoo) [inline addons/stock/product.py]: Unuseful dict creation... you will crush it at line 416

File: addons/stock/product.py [hunk 5]
```diff
@@ -402,6 +406,18 @@ def action_view_routes(self, cr, uid, ids, context=None):
...
+        if ids and track_all:
+            quant_ids = self.pool.get('stock.quant').search(cr, uid, [('product_id.product_tmpl_id','in',ids), ('lot_id','=', False), ('location_id.usage','=', 'internal')])
+            if quant_ids:
+                warning_msgs =  _("There are Quants without lot for this product. You need to assign serial number (e.g. by creating an inventory) first to prevent inventory loss.")
+                if warning_msgs:
```
Comment (rim-odoo) [inline addons/stock/product.py]: What is the probability that a string you defined on the previous line is void?

File: addons/stock/product.py [hunk 6]
```diff
@@ -402,6 +406,18 @@ def action_view_routes(self, cr, uid, ids, context=None):
...
+                    warning = {
+                        'title': _('Configuration Error!'),
+                        'message' : warning_msgs
+                       }
+        return {'warning' : warning}
```
Comment (rim-odoo) [inline addons/stock/product.py]: What if `warning` is empty?

File: addons/stock/product.py (hunk 3 above)
Comment (rim-odoo) [inline addons/stock/product.py]: Prefer `self.pool["my.model"]` to `self.pool.get("my.model")`, it will avoid some obscure crashes

Comment (rim-odoo) [inline addons/stock/product.py]: Missing `context` parameter

**Comment by mart-e** [comment]
Thanks @rim-odoo for the review.
Merged with corrections in master at e03ec16246f9e3a205e6c124ebf1b076529d1bf1

**Comment by rim-odoo** [comment]
my pleasure @mart-e :postal_horn:

**Pull request closed without merging**